api/src/main/java/org/apache/iceberg/expressions/Literals.java

```diff
@@ -426,16 +428,40 @@ protected Type.TypeID typeId() {
   }
 
   static class TimestampLiteral extends ComparableLiteral<Long> {
-    TimestampLiteral(Long value) {
+    private final ChronoUnit unit;
+
+    TimestampLiteral(ChronoUnit unit, Long value) {
       super(value);
+      this.unit = unit;
     }
 
     @Override
     @SuppressWarnings("unchecked")
     public <T> Literal<T> to(Type type) {
       switch (type.typeId()) {
         case TIMESTAMP:
-          return (Literal<T>) this;
+          ChronoUnit toUnit = ((TimestampType) type).unit();
+          switch (unit) {
+            case MICROS:
+              switch (toUnit) {
+                case MICROS:
+                  return (Literal<T>) this;
+                case NANOS:
+                  return (Literal<T>)
+                      new TimestampLiteral(unit, DateTimeUtil.microsToNanos(value()));
+              }
+              break;
+            case NANOS:
+              switch (toUnit) {
+                case MICROS:
+                  return (Literal<T>)
+                      new TimestampLiteral(unit, DateTimeUtil.nanosToMicros(value()));
+                case NANOS:
+                  return (Literal<T>) this;
+              }
+              break;
+          }
+          break;
```

Comment:
Conversion to date needs to be updated as well.

Reply:
I'll add that.

Reply:
Done.